README.md
# NAME

B::Tap - Inject tapping node to optree

# SYNOPSIS

    use B;
    use B::Tap;
    use B::Tools;

    sub foo { 63 }

    my $code = sub { foo() + 5900 };
    my $cv = B::svref_2object($code);

    my ($entersub) = op_grep { $_->name eq 'entersub' } $cv->ROOT;
    tap($$entersub, ${$cv->ROOT}, \my @buf);

    $code->();

# DESCRIPTION

B::Tap is tapping library for B tree. `tap` function injects custom ops for fetching result of the node.

The implementation works, but it's not beautiful code. I'm newbie about the B world, Patches welcome.

__WARNINGS: This module is in a alpha state. Any API will change without notice.__

# FUNCTIONS

- tap($op, $root\_op, \\@buf)

    Tapping the result value of `$op`. You need pass the `$root_op` for rewriting tree structure. Tapped result value was stored to `\@buf`. `\@buf` must be arrayref.

    B::Tap push the current stack to `\@buf`. First element for each value is `GIMME_V`. Second element is the value of stacks.

- G\_SCALAR
- G\_ARRAY
- G\_VOID

    These functions are not exportable by default. If you want to use these functions, specify the import arguments like:

        use B::Tap ':all';

    Or

        use B::Tap qw(G_SCALAR G_ARRAY G_VOID);

# FAQ

- Why this module required 5.14+?

    Under 5.14, Perl5's custom op support is incomplete. B::Deparse can't deparse the code using custom ops.

    I seem this library without deparsing is useless.

    But if you want to use this with 5.8, it may works.

# LICENSE

Copyright (C) tokuhirom.

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself.

# AUTHOR

tokuhirom <tokuhirom@gmail.com>


Changes
Revision history for Perl extension B-Tap

{{$NEXT}}

0.16 2020-07-10T06:54:42Z

    - Follow perl 5.22.0 op_sibling change.
      https://perldoc.perl.org/perl5220delta.html#Internal-Changes


0.15 2015-05-02T02:38:11Z

    - Skip tests on bled perl

0.14 2014-08-20T02:18:53Z

    commit 595a364add38356078177753b00574f4078ae7f8
    Author: Brian Fraser <fraserbn@gmail.com>
    Date:   Sat Aug 16 04:35:54 2014 +0200

        Avoid assertion failures in debugging perls

0.13 2014-03-19T11:05:53Z

    - Suppress warnings from Devel::CodeObserver

0.12 2014-03-14T01:42:23Z

    - [Devel::CodeObserver] Load B::Deparse lazily.
      (tokuhirom)

0.11 2014-03-10T01:31:58Z

    [INCOMPATIBLE CHANGES]
    - Devel::CodeObserver returns Devel::CodeObserver::Result instead of plain ArrayRef
      (tokuhirom)

0.10 2014-03-10T01:13:04Z

    - Added Devel::CodeObserver(formely known as Test::Power::Core)
      (tokuhirom)

0.08 2013-11-05T07:54:43Z

    [BUG FIX]
    - Fixed infinite loop bug.
      (tokuhirom)

0.07 2013-11-05T04:11:43Z

    [BUG FIX]
    - 0.06 does not works with ithreads.
      (tokuhirom)

0.06 2013-11-02T02:50:45Z

    - Care extended ops.
      (tokuhirom)

0.05 2013-09-18T04:54:19Z

    - Apply OPf_KIDS for b_tap_tap.
      (tokuhirom)

0.04 2013-09-15T01:24:08Z

    - oops.

0.03 2013-09-15T01:15:30Z

    - restore MARK on b_tap_tap

0.02 2013-09-14T06:59:09Z

    - Oops. I forgot to restore MARK.
      It fixed segmentation fault on 32bit machines.
      (Reported by syohex++)

0.01 2013-09-14T03:37:15Z

    - original version



Build.PL
# =========================================================================
# THIS FILE IS AUTOMATICALLY GENERATED BY MINILLA.
# DO NOT EDIT DIRECTLY.
# =========================================================================

use 5.008_001;

use strict;
use warnings;
use utf8;

use Module::Build;
use File::Basename;
use File::Spec;
use CPAN::Meta;
use CPAN::Meta::Prereqs;

my %args = (
    license              => 'perl',
    dynamic_config       => 0,

    configure_requires => {
        'Module::Build' => 0.38,
    },

    name            => 'B-Tap',
    module_name     => 'B::Tap',
    allow_pureperl => 0,

    script_files => [glob('script/*'), glob('bin/*')],
    c_source     => [qw()],
    PL_files => {},

    test_files           => ((-d '.git' || $ENV{RELEASE_TESTING}) && -d 'xt') ? 't/ xt/' : 't/',
    recursive_test_files => 1,

    
);
if (-d 'share') {
    $args{share_dir} = 'share';
}

my $builder = Module::Build->subclass(
    class => 'MyBuilder',
    code => q{
        sub ACTION_distmeta {
            die "Do not run distmeta. Install Minilla and `minil install` instead.\n";
        }
        sub ACTION_installdeps {
            die "Do not run installdeps. Run `cpanm --installdeps .` instead.\n";
        }
    }
)->new(%args);
$builder->create_build_script();

my $mbmeta = CPAN::Meta->load_file('MYMETA.json');
my $meta = CPAN::Meta->load_file('META.json');
my $prereqs_hash = CPAN::Meta::Prereqs->new(
    $meta->prereqs
)->with_merged_prereqs(
    CPAN::Meta::Prereqs->new($mbmeta->prereqs)
)->as_string_hash;
my $mymeta = CPAN::Meta->new(
    {
        %{$meta->as_struct},
        prereqs => $prereqs_hash
    }
);
print "Merging cpanfile prereqs to MYMETA.yml\n";
$mymeta->save('MYMETA.yml', { version => 1.4 });
print "Merging cpanfile prereqs to MYMETA.json\n";
$mymeta->save('MYMETA.json', { version => 2 });


lib/Devel/CallTrace.pm
package Devel::CallTrace;
use strict;
use warnings;
use utf8;
use 5.010_001;

our $VERSION = "0.09";

our @WARNINGS;

use B qw(class ppname);
use B::Tap qw(tap);
use B::Tools qw(op_walk);
use B::Deparse;
use Data::Dumper ();
use Try::Tiny;

sub new {
    my $class = shift;
    my %args = @_==1 ? %{$_[0]} : @_;
    bless {%args}, $class;
}

sub null {
    my $op = shift;
    return class($op) eq "NULL";
}

sub call {
    my ($class,$code) = @_;

    my $cv = B::svref_2object($code);

    my @tap_results;

    my $root = $cv->ROOT;
    # local $B::overlay = {};
    if (not null $root) {
        op_walk {
            if (need_hook($_)) {
                my @buf = ($_);
                tap($_, $cv->ROOT, \@buf);
                push @tap_results, \@buf;
            }
        } $cv->ROOT;
    }
    if (0) {
        require B::Concise;
        my $walker = B::Concise::compile('', '', $code);
        $walker->();
    }

    my $retval = $code->();

    return (
        $retval,
        dump_pairs($code, [grep { @$_ > 1 } @tap_results]),
    );
}

sub dump_pairs {
    my ($code, $tap_results) = @_;

    my @pairs;
    local $Data::Dumper::Terse = 1;
    local $Data::Dumper::Indent = 0;
    for my $result (@$tap_results) {
        my $op = shift @$result;
        for my $value (@$result) {
            # take first argument if the value is scalar.
            try {
                # Suppress warnings for: sub { expect(\@p)->to_be(['a']) }
                local $SIG{__WARN__} = sub { };

                my $deparse = B::Deparse->new();
                $deparse->{curcv} = B::svref_2object($code);
                push @pairs, $deparse->deparse($op);
                push @pairs, Data::Dumper::Dumper($value->[1]);
            } catch {
                warn "[Test::Power] [BUG]: $_";
                push @WARNINGS, "[Test::Power] [BUG]: $_";
            };
        }
    }
    return \@pairs;
}

sub need_hook {
    my $op = shift;
    return 1 if $op->name eq 'entersub';
    return 1 if $op->name eq 'padsv';
    return 1 if $op->name eq 'aelem';
    return 1 if $op->name eq 'helem';
    return 1 if $op->name eq 'null' && ppname($op->targ) eq 'pp_rv2sv';
    return 0;
}


1;
__END__

=head1 NAME

Devel::CallTrace - Code tracer

=head1 SYNOPSIS

    my $tracer = Devel::CallTrace->new();
    my ($retval, $trace_data) = $tracer->call(sub { $dat->{foo}{bar} eq 200 });

=head1 DESCRIPTION

This module call the CodeRef, and fetch the Perl5 VM's temporary values.

=head1 METHODS

=over 4

=item C<< my $tracer = Devel::CallTrace->new(); >>

Create new instance.

=item C<< $tracer->call($code: CodeRef) : (Scalar, ArrayRef) >>

Call the C<$code> and get the tracing result.

=back

=head1 EXAMPLES

Here is the concrete example.

    use 5.014000;
    use Devel::CallTrace;
    use Data::Dumper;

    my $dat = {
        x => {
            y => 0,
        },
        z => {
            m => [
                +{
                    n => 3
                }
            ]
        }
    };

    my $tracer = Devel::CallTrace->new();
    my ($retval, $traced) = $tracer->call(sub { $dat->{z}->{m}[0]{n} eq 4 ? 1 : 0 });
    print "RETVAL: $retval\n";
    while (my ($code, $val) = splice @$traced, 0, 2) {
        print "$code => $val\n";
    }

Output is here:

    RETVAL: 0
    $$dat{'z'}{'m'}[0]{'n'} => 3
    $$dat{'z'}{'m'}[0] => {'n' => 3}
    $$dat{'z'}{'m'} => [{'n' => 3}]
    $$dat{'z'} => {'m' => [{'n' => 3}]}
    $dat => {'z' => {'m' => [{'n' => 3}]},'x' => {'y' => 0}}

Devel::CallTrace fetches the temporary values and return it.

=head1 BUGS

=head2 LIST CONTEXT

There is no list context support. I don't want to implement this, for now.
But you can send me a patch.

=head2 METHOD CALL

This version can't handles following form:

    my $tracer = Devel::CallTrace->new();
    $tracer->call(sub { defined($foo->bar()) });

Because B::Deparse::pp_entersub thinks next object is the `method_named` or LISTOP.
But B::Tap's b_tap_push_sv is SVOP!!!

I should fix this issue, but I have no time to fix this.

Patches welcome.



lib/Devel/CodeObserver.pm
package Devel::CodeObserver;
use strict;
use warnings;
use utf8;
use 5.010_001;

our $VERSION = "0.14";

our @WARNINGS;

use B qw(class ppname);
use B::Tap qw(tap);
use B::Tools qw(op_walk);
use Data::Dumper ();

sub new {
    my $class = shift;
    my %args = @_==1 ? %{$_[0]} : @_;
    bless {%args}, $class;
}

sub null {
    my $op = shift;
    return class($op) eq "NULL";
}

sub call {
    my ($class,$code) = @_;

    my $cv = B::svref_2object($code);

    my @tap_results;

    my $root = $cv->ROOT;
    # local $B::overlay = {};
    if (not null $root) {
        op_walk {
            if (need_hook($_)) {
                my @buf = ($_);
                tap($_, $cv->ROOT, \@buf);
                push @tap_results, \@buf;
            }
        } $cv->ROOT;
    }
    if (0) {
        require B::Concise;
        my $walker = B::Concise::compile('', '', $code);
        $walker->();
    }

    my $retval = $code->();

    return (
        $retval,
        Devel::CallTrace::Result->new(
            code => $code,
            tap_results => [grep { @$_ > 1 } @tap_results],
        )
    );
}

sub need_hook {
    my $op = shift;
    return 1 if $op->name eq 'entersub';
    return 1 if $op->name eq 'padsv';
    return 1 if $op->name eq 'aelem';
    return 1 if $op->name eq 'helem';
    return 1 if $op->name eq 'null' && ppname($op->targ) eq 'pp_rv2sv';
    return 0;
}

package Devel::CallTrace::Result;

use Try::Tiny;
use constant { DEBUG => 0 };

sub new {
    my $class = shift;
    my %args = @_==1 ? %{$_[0]} : @_;
    bless {%args}, $class;
}

sub dump_pairs {
    my ($self) = @_;
    my $tap_results = $self->{tap_results};
    my $code = $self->{code};

    # We should load B::Deparse lazily. Because loading B::Deparse is really slow.
    # It's really big module.
    #
    # And so, this module is mainly used for testing. And this part is only required if
    # the test case was failed. I make faster the passed test case.
    require B::Deparse;

    my @pairs;
    local $Data::Dumper::Terse = 1;
    local $Data::Dumper::Indent = 0;
    for my $result (@$tap_results) {
        my $op = shift @$result;
        for my $value (@$result) {
            # take first argument if the value is scalar.
            try {
                # Suppress warnings for: sub { expect(\@p)->to_be(['a']) }
                local $SIG{__WARN__} = sub { };

                my $deparse = B::Deparse->new();
                $deparse->{curcv} = B::svref_2object($code);
                push @pairs, [
                    $deparse->deparse($op),
                    Data::Dumper::Dumper($value->[1])
                ];
            } catch {
                DEBUG && warn "[Devel::CodeObserver] [BUG]: $_";
                push @WARNINGS, "[Devel::CodeObserver] [BUG]: $_";
            };
        }
    }
    return \@pairs;
}

1;
__END__

=head1 NAME

Devel::CodeObserver - Code tracer

=head1 SYNOPSIS

    my $tracer = Devel::CodeObserver->new();
    my ($retval, $trace_data) = $tracer->call(sub { $dat->{foo}{bar} eq 200 });

=head1 DESCRIPTION

This module call the CodeRef, and fetch the Perl5 VM's temporary values.

=head1 METHODS

=over 4

=item C<< my $tracer = Devel::CodeObserver->new(); >>

Create new instance.

=item C<< $tracer->call($code: CodeRef) : (Scalar, Devel::CodeObserver::Result) >>

Call the C<$code> and get the tracing result.

=back

=head1 Devel::CodeObserver::Result's METHODS

=over 4

=item C<< $result->dump_pairs() : ArrayRef[ArrayRef[Str]] >>

Returns the pair of the dump result. Return value's each element contains ArrayRef.
Each element contains 2 values. First is the B::Deparse'd code. Second is the Dumper()'ed value.

=back

=head1 EXAMPLES

Here is the concrete example.

    use 5.014000;
    use Devel::CodeObserver;
    use Data::Dumper;

    my $dat = {
        x => {
            y => 0,
        },
        z => {
            m => [
                +{
                    n => 3
                }
            ]
        }
    };

    my $tracer = Devel::CodeObserver->new();
    my ($retval, $result) = $tracer->call(sub { $dat->{z}->{m}[0]{n} eq 4 ? 1 : 0 });
    print "RETVAL: $retval\n";
    for my $pair (@{$result->dump_pairs}) {
        my ($code, $val) = @$pair;
        print "$code => $val\n";
    }

Output is here:

    RETVAL: 0
    $$dat{'z'}{'m'}[0]{'n'} => 3
    $$dat{'z'}{'m'}[0] => {'n' => 3}
    $$dat{'z'}{'m'} => [{'n' => 3}]
    $$dat{'z'} => {'m' => [{'n' => 3}]}
    $dat => {'z' => {'m' => [{'n' => 3}]},'x' => {'y' => 0}}

Devel::CodeObserver fetches the temporary values and return it.

=head1 BUGS

=head2 LIST CONTEXT

There is no list context support. I don't want to implement this, for now.
But you can send me a patch.

=head2 METHOD CALL

This version can't handles following form:

    my $tracer = Devel::CodeObserver->new();
    $tracer->call(sub { defined($foo->bar()) });

Because B::Deparse::pp_entersub thinks next object is the `method_named` or LISTOP.
But B::Tap's b_tap_push_sv is SVOP!!!

I should fix this issue, but I have no time to fix this.

Patches welcome.



lib/B/Tap.pm
package B::Tap;
use 5.014000;
use strict;
use warnings;

our $VERSION = "0.14";

use parent qw(Exporter);

our @EXPORT = qw(tap);
our @EXPORT_OK = qw(G_ARRAY G_VOID G_SCALAR);
our %EXPORT_TAGS = (
    'all' => [@EXPORT, @EXPORT_OK],
);

use Carp ();

use XSLoader;
XSLoader::load(__PACKAGE__, $VERSION);

sub tap {
    my ($op, $root_op, $buf) = @_;
    Carp::croak("Third argument should be ArrayRef") unless ref $buf eq 'ARRAY';
    _tap($$op, $$root_op, $buf);
}

# tweaks for custom ops.
{
    sub B::Deparse::pp_b_tap_tap {
        my ($self, $op) = @_;
        $self->deparse($op->first);
    };
    sub B::Deparse::pp_b_tap_push_sv {
        '';
    }
}

1;
__END__

=for stopwords optree newbie deparse deparsing

=encoding utf-8

=head1 NAME

B::Tap - Inject tapping node to optree

=head1 SYNOPSIS

    use B;
    use B::Tap;
    use B::Tools;

    sub foo { 63 }

    my $code = sub { foo() + 5900 };
    my $cv = B::svref_2object($code);

    my ($entersub) = op_grep { $_->name eq 'entersub' } $cv->ROOT;
    tap($$entersub, ${$cv->ROOT}, \my @buf);

    $code->();

=head1 DESCRIPTION

B::Tap is tapping library for B tree. C<tap> function injects custom ops for fetching result of the node.

The implementation works, but it's not beautiful code. I'm newbie about the B world, Patches welcome.

B<WARNINGS: This module is in a alpha state. Any API will change without notice.>

=head1 FUNCTIONS

=over 4

=item tap($op, $root_op, \@buf)

Tapping the result value of C<$op>. You need pass the C<$root_op> for rewriting tree structure. Tapped result value was stored to C<\@buf>. C<\@buf> must be arrayref.

B::Tap push the current stack to C<\@buf>. First element for each value is C<GIMME_V>. Second element is the value of stacks.

=item G_SCALAR

=item G_ARRAY

=item G_VOID

These functions are not exportable by default. If you want to use these functions, specify the import arguments like:

    use B::Tap ':all';

Or

    use B::Tap qw(G_SCALAR G_ARRAY G_VOID);

=back

=head1 FAQ

=over 4

=item Why this module required 5.14+?

Under 5.14, Perl5's custom op support is incomplete. B::Deparse can't deparse the code using custom ops.

I seem this library without deparsing is useless.

But if you want to use this with 5.8, it may works.

=back

=head1 LICENSE

Copyright (C) tokuhirom.

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself.

=head1 AUTHOR

tokuhirom E<lt>tokuhirom@gmail.comE<gt>

=cut

